CodingGuidelines.md

```diff
@@ -0,0 +1,168 @@
+# Coding Guidelines
+
+This document outlines the recommended coding guidelines for the Mixed Reality Toolkit.  The majority of these suggestions follow the [recommended standards from MSDN](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/inside-a-program/coding-conventions).
+
+## Spaces vs Tabs
+Please be sure to use 4 spaces when contributing to this project.
+
+## Naming Conventions
+
+Always use `PascalCase` for public properties, and `camelCase` for private properties and fields.
+> The only exception to this is for data structures that require the fields to be serialized by the `JsonUtility`.
+
+## Access Modifiers
+
+Always declare an access modifier for all fields, properties and methods.
+
+>All Unity API Methods should be `private` by default, unless you need to override them in a derived class. In this case `protected` should be used.
+
+### Don't:
+
+```
+void Foo(){ }
+void Bar(){ }
+```
+
+### Do:
+
+ ```
+private void Foo(){ }
+public void Bar(){ }
+ ```
+
+## Use Braces
+Always use braces whenever possible.
+
+### Don't:
+
+```
+private Foo()
+{
+    if(Bar==null)
+        DoThing();
+    else
+        DoTheOtherThing();
+}
+```
+
+### Do:
+
+```
+private Foo()
+{
+    if(Bar==true)
+    {
+        DoThing();
+    }
+    else
+    {
+        DoTheOtherThing();
+    }
+}
+```
+
+## Encapsulation
+Always use private fields and public properties if access to the field is needed from outside the class or struct.
+
+If you need to have the ability to edit your field in the inspector, it's best practice to follow the rules for Encapsulation and serialize your backing field.
+
+>The only exception to this is for data structures that require the fields to be serialized by the `JsonUtility`.
+
+### Don't:
+
+```
+public float MyValue;
+```
+
+### Do:
+
+ ```
+ [SerializeField] // Only use this attribute to view/edit in inspector.
+ private float myValue;
+ public float MyValue
+ {
+     get{ return myValue; }
+     set{ myValue = value }
+ }
+ ```
+
+## Best Practices
+Some of the target platforms of this project require us to take performance into consideration.  What this in mind we should always be careful of allocating memory in frequently called code in tight update loops or algorithms.
+
+---
+
+Use `for` instead of `foreach` when possible.
```

Comment:
When would it not be possible to use a for? all enumerable collections support a Size/Count property, to the best of my recollection.
 If there are situations where a foreach is preferred (or acceptable) we should document an exception.